File: web/src/api/agentAdmin.ts
/** Agent 运维只读 API（与 `agent_admin.api_keys` 一致，Bearer / X-API-Key） */

function parseAdminError(text: string, status: number): string {
  let msg = `HTTP ${status}`
  if (!text) return msg
  try {
    const j = JSON.parse(text) as {
      code?: number
      reason?: string
      message?: string
      error?: { message?: string; code?: string }
    }
    if (typeof j.error?.message === 'string' && j.error.message) {
      msg = j.error.message
    } else {
      const reason = typeof j.reason === 'string' ? j.reason : ''
      const message = typeof j.message === 'string' ? j.message : ''
      if (reason && message && reason !== message) {
        msg = `${reason}: ${message}`
      } else if (message) {
        msg = message
      } else if (reason) {
        msg = reason
      } else {
        msg = text.slice(0, 300)
      }
    }
  } catch {
    msg = text.slice(0, 300)
  }
  return msg
}

async function readJson<T>(res: Response): Promise<T> {
  const text = await res.text()
  if (!res.ok) {
    throw new Error(parseAdminError(text, res.status))
  }
  if (!text) return {} as T
  return JSON.parse(text) as T
}

function authHeaders(apiKey: string): HeadersInit {
  return { Authorization: `Bearer ${apiKey.trim()}` }
}

/** 与 Kratos/proto JSON 一致（camelCase） */
export interface AgentSummary {
  agentId: string
  queue: string
  hostname: string
  /** protobuf Timestamp JSON：RFC3339 */
  lastTaskHeartbeatAt?: string
  online: boolean
  /** online | offline | unknown（无 last_task_heartbeat_at） */
  status?: string
}

export interface ListAgentsReply {
  agents: AgentSummary[]
  /** 离线判定窗口（秒），与 agent.offline_min_sec 等一致 */
  offlineWindowSec?: number
  offline_window_sec?: number
}

export async function listAgents(apiKey: string): Promise<ListAgentsReply> {
  const res = await fetch('/api/v1/admin/agents', {
    headers: authHeaders(apiKey),
  })
  return readJson<ListAgentsReply>(res)
}

export interface LeasedTaskRow {
  taskId: string
  scriptId: string
  version: string
  leasedAt?: string
  leaseDeadlineAt?: string
}

export interface ListAgentLeasedTasksReply {
  tasks: LeasedTaskRow[]
}

export async function listAgentLeasedTasks(apiKey: string, agentId: string): Promise<ListAgentLeasedTasksReply> {
  const id = encodeURIComponent(agentId.trim())
  const res = await fetch(`/api/v1/admin/agents/${id}/leased-tasks`, {
    headers: authHeaders(apiKey),
  })
  return readJson<ListAgentLeasedTasksReply>(res)
}

export interface InstalledScriptRow {
  scriptId: string
  version: string
  envStatus: string
  updatedAt?: string
}

export interface ListAgentInstalledScriptsReply {
  scripts: InstalledScriptRow[]
}

export async function listAgentInstalledScripts(
  apiKey: string,
  agentId: string,
): Promise<ListAgentInstalledScriptsReply> {
  const id = encodeURIComponent(agentId.trim())
  const res = await fetch(`/api/v1/admin/agents/${id}/installed-scripts`, {
    headers: authHeaders(apiKey),
  })
  return readJson<ListAgentInstalledScriptsReply>(res)
}

/** 平台登录凭据（密码仅 Upsert 时提交，列表不返回密码） */
export interface AgentScriptAuthRow {
  scriptId: string
  username: string
  updatedAt?: string
}

export interface ListAgentScriptAuthsReply {
  rows?: AgentScriptAuthRow[]
}

function normalizeScriptAuthRows(raw: ListAgentScriptAuthsReply): AgentScriptAuthRow[] {
  const rows = raw.rows
  if (!Array.isArray(rows)) return []
  return rows.map((r) => {
    const o = r as unknown as Record<string, unknown>
    const scriptId =
      typeof o.scriptId === 'string'
        ? o.scriptId
        : typeof o.script_id === 'string'
          ? o.script_id
          : ''
    const username =
      typeof o.username === 'string'
        ? o.username
        : typeof o.userName === 'string'
          ? o.userName
          : ''
    let updatedAt: string | undefined
    const u = o.updatedAt ?? o.updated_at
    if (typeof u === 'string') updatedAt = u
    else if (u && typeof u === 'object' && 'seconds' in u) {
      const sec = (u as { seconds?: string | number }).seconds
      const nano = (u as { nanos?: number }).nanos ?? 0
      if (sec != null) {
        const t = Number(sec) * 1000 + Math.floor(nano / 1e6)
        updatedAt = new Date(t).toISOString()
      }
    }
    return { scriptId, username, updatedAt }
  })
}

export async function listAgentScriptAuths(
  apiKey: string,
  agentId: string,
): Promise<{ rows: AgentScriptAuthRow[] }> {
  const id = encodeURIComponent(agentId.trim())
  const res = await fetch(`/api/v1/admin/agents/${id}/script-auths`, {
    headers: authHeaders(apiKey),
  })
  const raw = await readJson<ListAgentScriptAuthsReply>(res)
  return { rows: normalizeScriptAuthRows(raw) }
}

export async function upsertAgentScriptAuth(
  apiKey: string,
  params: { agentId: string; scriptId: string; username: string; password: string },
): Promise<void> {
  const aid = encodeURIComponent(params.agentId.trim())
  const sid = encodeURIComponent(params.scriptId.trim())
  // Kratos 对 proto 请求体使用 protojson：字段名须为 camelCase（agentId/scriptId），snake_case 会被 DiscardUnknown 丢掉
  const res = await fetch(`/api/v1/admin/agents/${aid}/script-auths/${sid}`, {
    method: 'PUT',
    headers: { ...authHeaders(apiKey), 'Content-Type': 'application/json' },
    body: JSON.stringify({
      agentId: params.agentId.trim(),
      scriptId: params.scriptId.trim(),
      username: params.username.trim(),
      password: params.password,
    }),
  })
  await readJson<Record<string, unknown>>(res)
}

export async function deleteAgentScriptAuth(
  apiKey: string,
  agentId: string,
  scriptId: string,
): Promise<void> {
  const aid = encodeURIComponent(agentId.trim())
  const sid = encodeURIComponent(scriptId.trim())
  const res = await fetch(`/api/v1/admin/agents/${aid}/script-auths/${sid}`, {
    method: 'DELETE',
    headers: authHeaders(apiKey),
  })
  await readJson<Record<string, unknown>>(res)
}

// ---------- BOM 采集平台 /api/v1/admin/bom-platforms（同 agent_admin 鉴权）----------

function unwrapProtoValue(v: unknown): unknown {
  if (v === null || typeof v !== 'object' || Array.isArray(v)) return v
  const o = v as Record<string, unknown>
  if ('numberValue' in o) return o.numberValue
  if ('stringValue' in o) return o.stringValue
  if ('boolValue' in o) return o.boolValue
  if ('nullValue' in o) return null
  return v
}

/** 将 API 中的 Struct / 平铺 JSON 转为可编辑的 plain 对象 */
export function runParamsToPlain(input: unknown): Record<string, unknown> | null {
  if (input == null) return null
  if (typeof input !== 'object' || Array.isArray(input)) return null
  const o = input as Record<string, unknown>
  const fields = o.fields
  if (fields && typeof fields === 'object' && !Array.isArray(fields)) {
    const out: Record<string, unknown> = {}
    for (const [k, val] of Object.entries(fields as Record<string, unknown>)) {
      out[k] = unwrapProtoValue(val)
    }
    return out
  }
  return { ...o }
}

export interface BomPlatformRow {
  platformId: string
  scriptId: string
  displayName: string
  enabled: boolean
  runParams: Record<string, unknown> | null
  updatedAt?: string
}

function pickStr(o: Record<string, unknown>, camel: string, snake: string): string {
  const a = o[camel]
  const b = o[snake]
  if (typeof a === 'string') return a
  if (typeof b === 'string') return b
  return ''
}

function tsToIso(v: unknown): string | undefined {
  if (typeof v === 'string') return v
  if (v && typeof v === 'object' && 'seconds' in (v as object)) {
    const sec = (v as { seconds?: string | number }).seconds
    const nano = (v as { nanos?: number }).nanos ?? 0
    if (sec != null) {
      const t = Number(sec) * 1000 + Math.floor(nano / 1e6)
      return new Date(t).toISOString()
    }
  }
  return undefined
}

function normalizeBomPlatformRow(raw: Record<string, unknown>): BomPlatformRow {
  const runParamsRaw = raw.runParams ?? raw.run_params
  return {
    platformId: pickStr(raw, 'platformId', 'platform_id'),
    scriptId: pickStr(raw, 'scriptId', 'script_id'),
    displayName: pickStr(raw, 'displayName', 'display_name'),
    enabled: raw.enabled !== false && raw.enabled !== 'false' && raw.enabled !== 0,
    runParams: runParamsToPlain(runParamsRaw),
    updatedAt: tsToIso(raw.updatedAt ?? raw.updated_at),
  }
}

export interface ListBomPlatformsReply {
  items: BomPlatformRow[]
}

export async function listBomPlatforms(apiKey: string): Promise<ListBomPlatformsReply> {
  const res = await fetch('/api/v1/admin/bom-platforms', {
    headers: authHeaders(apiKey),
  })
  const raw = await readJson<{ items?: unknown[] }>(res)
  const items = Array.isArray(raw.items) ? raw.items : []
  return {
    items: items.map((it) => normalizeBomPlatformRow(it as Record<string, unknown>)),
  }
}

export async function getBomPlatform(apiKey: string, platformId: string): Promise<BomPlatformRow | null> {
  const id = encodeURIComponent(platformId.trim())
  const res = await fetch(`/api/v1/admin/bom-platforms/${id}`, {
    headers: authHeaders(apiKey),
  })
  const raw = await readJson<{ item?: Record<string, unknown> }>(res)
  const item = raw.item
  if (!item || typeof item !== 'object') return null
  return normalizeBomPlatformRow(item)
}

export async function upsertBomPlatform(
  apiKey: string,
  body: {
    platformId: string
    scriptId: string
    displayName: string
    enabled: boolean
    /** 不传或 undefined：请求体不含 runParams，后端将 run_params 置空（NULL）；传对象则全量写入（可为 {}） */
    runParams?: Record<string, unknown>
  },
): Promise<BomPlatformRow> {
  const id = encodeURIComponent(body.platformId.trim())
  const payload: Record<string, unknown> = {
    platformId: body.platformId.trim(),
    scriptId: body.scriptId.trim(),
    displayName: body.displayName.trim(),
    enabled: body.enabled,
  }
  if (body.runParams !== undefined) {
    payload.runParams = body.runParams
  }
  const res = await fetch(`/api/v1/admin/bom-platforms/${id}`, {
    method: 'PUT',
    headers: { ...authHeaders(apiKey), 'Content-Type': 'application/json' },
    body: JSON.stringify(payload),
  })
  const raw = await readJson<{ item?: Record<string, unknown> }>(res)
  const item = raw.item
  if (!item || typeof item !== 'object') {
    throw new Error('Upsert 成功但未返回 item')
  }
  return normalizeBomPlatformRow(item)
}

export async function deleteBomPlatform(apiKey: string, platformId: string): Promise<void> {
  const id = encodeURIComponent(platformId.trim())
  const res = await fetch(`/api/v1/admin/bom-platforms/${id}`, {
    method: 'DELETE',
    headers: authHeaders(apiKey),
  })
  await readJson<Record<string, unknown>>(res)
}

File: internal/service/agent_admin.go
package service

import (
	"context"
	"encoding/json"
	"strings"
	"time"

	v1 "caichip/api/admin/v1"
	"caichip/internal/biz"
	"caichip/internal/conf"

	kerrors "github.com/go-kratos/kratos/v2/errors"
	"github.com/go-kratos/kratos/v2/log"
	khttp "github.com/go-kratos/kratos/v2/transport/http"
	"google.golang.org/protobuf/types/known/structpb"
	"google.golang.org/protobuf/types/known/timestamppb"
)

// AgentAdminService 实现 api/admin/v1.AgentAdminServiceHTTPServer（与 Agent 采集 API 分离）。
type AgentAdminService struct {
	bc         *conf.Bootstrap
	registry   biz.AgentRegistryRepo
	dispatch   biz.DispatchTaskRepo
	scriptAuth biz.AgentScriptAuthRepo
	platforms  biz.BomPlatformScriptRepo
	log        *log.Helper
}

// NewAgentAdminService ...
func NewAgentAdminService(
	bc *conf.Bootstrap,
	reg biz.AgentRegistryRepo,
	disp biz.DispatchTaskRepo,
	scriptAuth biz.AgentScriptAuthRepo,
	platforms biz.BomPlatformScriptRepo,
	logger log.Logger,
) *AgentAdminService {
	return &AgentAdminService{
		bc: bc, registry: reg, dispatch: disp, scriptAuth: scriptAuth, platforms: platforms, log: log.NewHelper(logger),
	}
}

// Enabled 是否配置了 agent_admin.api_keys（用于注册 HTTP）。
func (s *AgentAdminService) Enabled() bool {
	if s == nil || s.bc == nil || s.bc.AgentAdmin == nil {
		return false
	}
	for _, k := range s.bc.AgentAdmin.ApiKeys {
		if strings.TrimSpace(k) != "" {
			return true
		}
	}
	return false
}

func (s *AgentAdminService) validateAdminKey(ctx context.Context) bool {
	if s == nil || s.bc == nil || s.bc.AgentAdmin == nil {
		return false
	}
	keys := make(map[string]struct{})
	for _, k := range s.bc.AgentAdmin.ApiKeys {
		k = strings.TrimSpace(k)
		if k != "" {
			keys[k] = struct{}{}
		}
	}
	if len(keys) == 0 {
		return false
	}
	r, ok := khttp.RequestFromServerContext(ctx)
	if !ok {
		return false
	}
	if x := strings.TrimSpace(r.Header.Get("X-API-Key")); x != "" {
		_, hit := keys[x]
		return hit
	}
	const p = "Bearer "
	auth := r.Header.Get("Authorization")
	if strings.HasPrefix(auth, p) {
		token := strings.TrimSpace(auth[len(p):])
		_, hit := keys[token]
		return hit
	}
	return false
}

func (s *AgentAdminService) dbOK() bool {
	return s.registry != nil && s.registry.DBOk() && s.dispatch != nil && s.dispatch.DBOk() &&
		s.scriptAuth != nil && s.scriptAuth.DBOk() && s.platforms != nil && s.platforms.DBOk()
}

func bomPlatformRowToProto(p *biz.BomPlatformScript) (*v1.BomPlatformRow, error) {
	if p == nil {
		return nil, nil
	}
	var st *structpb.Struct
	if len(p.RunParamsJSON) > 0 {
		var m map[string]interface{}
		if err := json.Unmarshal(p.RunParamsJSON, &m); err != nil {
			return nil, err
		}
		var err error
		st, err = structpb.NewStruct(m)
		if err != nil {
			return nil, err
		}
	}
	return &v1.BomPlatformRow{
		PlatformId:  p.PlatformID,
		ScriptId:    p.ScriptID,
		DisplayName: p.DisplayName,
		Enabled:     p.Enabled,
		RunParams:   st,
		UpdatedAt:   timestamppb.New(p.UpdatedAt),
	}, nil
}

// ListAgents 列出 Agent 及在线状态（与 BootstrapAgentOfflineThreshold 一致）。
func (s *AgentAdminService) ListAgents(ctx context.Context, _ *v1.ListAgentsRequest) (*v1.ListAgentsReply, error) {
	if !s.validateAdminKey(ctx) {
		return nil, kerrors.Unauthorized("UNAUTHORIZED", "invalid or missing agent admin api key")
	}
	if !s.dbOK() {
		return nil, kerrors.ServiceUnavailable("DB_DISABLED", "database not configured")
	}
	// BootstrapAgentOfflineThreshold 返回 time.Duration（配置里的秒已乘 time.Second）
	th := biz.BootstrapAgentOfflineThreshold(s.bc)
	now := time.Now()
	cutoff := now.Add(-th)
	if _, err := s.registry.MarkAgentsOfflineBefore(ctx, cutoff); err != nil {
		return nil, err
	}
	rows, err := s.registry.ListAgentRegistrySummaries(ctx)
	if err != nil {
		return nil, err
	}
	out := &v1.ListAgentsReply{
		Agents:           make([]*v1.AgentSummary, 0, len(rows)),
		OfflineWindowSec: int32(th / time.Second),
	}
	for _, r := range rows {
		online := false
		statusStr := biz.AgentStatusUnknown
		var ts *timestamppb.Timestamp
		if r.LastTaskHeartbeatAt != nil {
			ts = timestamppb.New(*r.LastTaskHeartbeatAt)
			if now.Sub(*r.LastTaskHeartbeatAt) <= th {
				online = true
				statusStr = biz.AgentStatusOnline
			} else {
				statusStr = biz.AgentStatusOffline
			}
		}
		out.Agents = append(out.Agents, &v1.AgentSummary{
			AgentId:             r.AgentID,
			Queue:               r.Queue,
			Hostname:            r.Hostname,
			LastTaskHeartbeatAt: ts,
			Online:              online,
			Status:              statusStr,
		})
	}
	return out, nil
}

// ListAgentLeasedTasks 某 Agent 当前租约中的调度任务。
func (s *AgentAdminService) ListAgentLeasedTasks(ctx context.Context, req *v1.ListAgentLeasedTasksRequest) (*v1.ListAgentLeasedTasksReply, error) {
	if !s.validateAdminKey(ctx) {
		return nil, kerrors.Unauthorized("UNAUTHORIZED", "invalid or missing agent admin api key")
	}
	if !s.dbOK() {
		return nil, kerrors.ServiceUnavailable("DB_DISABLED", "database not configured")
	}
	aid := strings.TrimSpace(req.GetAgentId())
	if aid == "" {
		return nil, kerrors.BadRequest("BAD_REQUEST", "agent_id required")
	}
	rows, err := s.dispatch.ListLeasedTasksByAgent(ctx, aid)
	if err != nil {
		return nil, err
	}
	out := &v1.ListAgentLeasedTasksReply{Tasks: make([]*v1.LeasedTaskRow, 0, len(rows))}
	for _, r := range rows {
		row := &v1.LeasedTaskRow{
			TaskId:   r.TaskID,
			ScriptId: r.ScriptID,
			Version:  r.Version,
		}
		if r.LeasedAt != nil {
			row.LeasedAt = timestamppb.New(*r.LeasedAt)
		}
		if r.LeaseDeadlineAt != nil {
			row.LeaseDeadlineAt = timestamppb.New(*r.LeaseDeadlineAt)
		}
		out.Tasks = append(out.Tasks, row)
	}
	return out, nil
}

// ListAgentInstalledScripts 某 Agent 已安装脚本（来自心跳上报快照）。
func (s *AgentAdminService) ListAgentInstalledScripts(ctx context.Context, req *v1.ListAgentInstalledScriptsRequest) (*v1.ListAgentInstalledScriptsReply, error) {
	if !s.validateAdminKey(ctx) {
		return nil, kerrors.Unauthorized("UNAUTHORIZED", "invalid or missing agent admin api key")
	}
	if !s.dbOK() {
		return nil, kerrors.ServiceUnavailable("DB_DISABLED", "database not configured")
	}
	aid := strings.TrimSpace(req.GetAgentId())
	if aid == "" {
		return nil, kerrors.BadRequest("BAD_REQUEST", "agent_id required")
	}
	rows, err := s.registry.ListInstalledScriptsForAgent(ctx, aid)
	if err != nil {
		return nil, err
	}
	out := &v1.ListAgentInstalledScriptsReply{Scripts: make([]*v1.InstalledScriptRow, 0, len(rows))}
	for _, r := range rows {
		out.Scripts = append(out.Scripts, &v1.InstalledScriptRow{
			ScriptId:  r.ScriptID,
			Version:   r.Version,
			EnvStatus: r.EnvStatus,
			UpdatedAt: timestamppb.New(r.UpdatedAt),
		})
	}
	return out, nil
}

// ListAgentScriptAuths 列出某 Agent 的 script 凭据（不含密码）。
func (s *AgentAdminService) ListAgentScriptAuths(ctx context.Context, req *v1.ListAgentScriptAuthsRequest) (*v1.ListAgentScriptAuthsReply, error) {
	if !s.validateAdminKey(ctx) {
		return nil, kerrors.Unauthorized("UNAUTHORIZED", "invalid or missing agent admin api key")
	}
	if !s.dbOK() {
		return nil, kerrors.ServiceUnavailable("DB_DISABLED", "database not configured")
	}
	aid := strings.TrimSpace(req.GetAgentId())
	if aid == "" {
		return nil, kerrors.BadRequest("BAD_REQUEST", "agent_id required")
	}
	rows, err := s.scriptAuth.ListByAgent(ctx, aid)
	if err != nil {
		return nil, err
	}
	out := &v1.ListAgentScriptAuthsReply{Rows: make([]*v1.AgentScriptAuthRow, 0, len(rows))}
	for _, r := range rows {
		out.Rows = append(out.Rows, &v1.AgentScriptAuthRow{
			ScriptId:  r.ScriptID,
			Username:  r.Username,
			UpdatedAt: timestamppb.New(r.UpdatedAt),
		})
	}
	return out, nil
}

// UpsertAgentScriptAuth 写入或更新凭据（须配置 AES 密钥）。
func (s *AgentAdminService) UpsertAgentScriptAuth(ctx context.Context, req *v1.UpsertAgentScriptAuthRequest) (*v1.UpsertAgentScriptAuthReply, error) {
	if !s.validateAdminKey(ctx) {
		return nil, kerrors.Unauthorized("UNAUTHORIZED", "invalid or missing agent admin api key")
	}
	if !s.dbOK() {
		return nil, kerrors.ServiceUnavailable("DB_DISABLED", "database not configured")
	}
	if s.scriptAuth == nil || !s.scriptAuth.CipherConfigured() {
		return nil, kerrors.BadRequest("SCRIPT_AUTH_CIPHER_DISABLED", "configure CAICHIP_AGENT_SCRIPT_AUTH_KEY or agent_script_auth.aes_key_base64 (32-byte key base64)")
	}
	aid := strings.TrimSpace(req.GetAgentId())
	sid := strings.TrimSpace(req.GetScriptId())
	if err := s.scriptAuth.Upsert(ctx, aid, sid, req.GetUsername(), req.GetPassword()); err != nil {
		return nil, kerrors.BadRequest("SCRIPT_AUTH_UPSERT", err.Error())
	}
	return &v1.UpsertAgentScriptAuthReply{}, nil
}

// DeleteAgentScriptAuth 删除一行凭据。
func (s *AgentAdminService) DeleteAgentScriptAuth(ctx context.Context, req *v1.DeleteAgentScriptAuthRequest) (*v1.DeleteAgentScriptAuthReply, error) {
	if !s.validateAdminKey(ctx) {
		return nil, kerrors.Unauthorized("UNAUTHORIZED", "invalid or missing agent admin api key")
	}
	if !s.dbOK() {
		return nil, kerrors.ServiceUnavailable("DB_DISABLED", "database not configured")
	}
	aid := strings.TrimSpace(req.GetAgentId())
	sid := strings.TrimSpace(req.GetScriptId())
	if err := s.scriptAuth.Delete(ctx, aid, sid); err != nil {
		return nil, err
	}
	return &v1.DeleteAgentScriptAuthReply{}, nil
}

// ListBomPlatforms 列出 BOM 采集平台及 run_params。
func (s *AgentAdminService) ListBomPlatforms(ctx context.Context, _ *v1.ListBomPlatformsRequest) (*v1.ListBomPlatformsReply, error) {
	if !s.validateAdminKey(ctx) {
		return nil, kerrors.Unauthorized("UNAUTHORIZED", "invalid or missing agent admin api key")
	}
	if !s.dbOK() {
		return nil, kerrors.ServiceUnavailable("DB_DISABLED", "database not configured")
	}
	rows, err := s.platforms.List(ctx)
	if err != nil {
		return nil, err
	}
	out := &v1.ListBomPlatformsReply{Items: make([]*v1.BomPlatformRow, 0, len(rows))}
	for i := range rows {
		row, err := bomPlatformRowToProto(&rows[i])
		if err != nil {
			return nil, kerrors.InternalServer("BOM_PLATFORM_SERIALIZE", err.Error())
		}
		out.Items = append(out.Items, row)
	}
	return out, nil
}

// GetBomPlatform 按 platform_id 查询一行。
func (s *AgentAdminService) GetBomPlatform(ctx context.Context, req *v1.GetBomPlatformRequest) (*v1.GetBomPlatformReply, error) {
	if !s.validateAdminKey(ctx) {
		return nil, kerrors.Unauthorized("UNAUTHORIZED", "invalid or missing agent admin api key")
	}
	if !s.dbOK() {
		return nil, kerrors.ServiceUnavailable("DB_DISABLED", "database not configured")
	}
	pid := strings.TrimSpace(req.GetPlatformId())
	if pid == "" {
		return nil, kerrors.BadRequest("BAD_REQUEST", "platform_id required")
	}
	row, err := s.platforms.Get(ctx, pid)
	if err != nil {
		return nil, err
	}
	if row == nil {
		return nil, kerrors.NotFound("BOM_PLATFORM_NOT_FOUND", "platform not found")
	}
	item, err := bomPlatformRowToProto(row)
	if err != nil {
		return nil, kerrors.InternalServer("BOM_PLATFORM_SERIALIZE", err.Error())
	}
	return &v1.GetBomPlatformReply{Item: item}, nil
}

// UpsertBomPlatform 创建或全量更新；要求 platform_id 与 script_id 一致。
func (s *AgentAdminService) UpsertBomPlatform(ctx context.Context, req *v1.UpsertBomPlatformRequest) (*v1.UpsertBomPlatformReply, error) {
	if !s.validateAdminKey(ctx) {
		return nil, kerrors.Unauthorized("UNAUTHORIZED", "invalid or missing agent admin api key")
	}
	if !s.dbOK() {
		return nil, kerrors.ServiceUnavailable("DB_DISABLED", "database not configured")
	}
	pid := strings.TrimSpace(req.GetPlatformId())
	sid := strings.TrimSpace(req.GetScriptId())
	if pid == "" || sid == "" {
		return nil, kerrors.BadRequest("BAD_REQUEST", "platform_id and script_id required")
	}
	if pid != sid {
		return nil, kerrors.BadRequest("BOM_PLATFORM_ID_MISMATCH", "platform_id must equal script_id")
	}
	var raw []byte
	if req.GetRunParams() != nil {
		m := req.GetRunParams().AsMap()
		b, err := json.Marshal(m)
		if err != nil {
			return nil, kerrors.BadRequest("RUN_PARAMS_JSON", err.Error())
		}
		raw = b
	}
	if _, err := biz.ExpandRunParamsJSON(raw); err != nil {
		return nil, kerrors.BadRequest("RUN_PARAMS_INVALID", err.Error())
	}
	row := &biz.BomPlatformScript{
		PlatformID:    pid,
		ScriptID:      sid,
		DisplayName:   strings.TrimSpace(req.GetDisplayName()),
		Enabled:       req.GetEnabled(),
		RunParamsJSON: raw,
	}
	if err := s.platforms.Upsert(ctx, row); err != nil {
		return nil, err
	}
	saved, err := s.platforms.Get(ctx, pid)
	if err != nil {
		return nil, err
	}
	if saved == nil {
		return nil, kerrors.InternalServer("BOM_PLATFORM_UPSERT", "row missing after upsert")
	}
	item, err := bomPlatformRowToProto(saved)
	if err != nil {
		return nil, kerrors.InternalServer("BOM_PLATFORM_SERIALIZE", err.Error())
	}
	return &v1.UpsertBomPlatformReply{Item: item}, nil
}

// DeleteBomPlatform 物理删除一行（停用请用 Upsert enabled=false）。
func (s *AgentAdminService) DeleteBomPlatform(ctx context.Context, req *v1.DeleteBomPlatformRequest) (*v1.DeleteBomPlatformReply, error) {
	if !s.validateAdminKey(ctx) {
		return nil, kerrors.Unauthorized("UNAUTHORIZED", "invalid or missing agent admin api key")
	}
	if !s.dbOK() {
		return nil, kerrors.ServiceUnavailable("DB_DISABLED", "database not configured")
	}
	pid := strings.TrimSpace(req.GetPlatformId())
	if pid == "" {
		return nil, kerrors.BadRequest("BAD_REQUEST", "platform_id required")
	}
	if err := s.platforms.Delete(ctx, pid); err != nil {
		return nil, err
	}
	return &v1.DeleteBomPlatformReply{}, nil
}

var _ v1.AgentAdminServiceHTTPServer = (*AgentAdminService)(nil)
